.ci/compute_projects.py

```diff
@@ -0,0 +1,186 @@
+        if modified_project not in PROJECT_CHECK_TARGETS:
+            continue
+        projects_to_test.add(modified_project)
+        if modified_project not in DEPENDENTS_TO_TEST:
+            continue
+        projects_to_test.update(DEPENDENTS_TO_TEST[modified_project])
+    if platform == "Linux":
+        for to_exclude in EXCLUDE_LINUX:
+            if to_exclude in projects_to_test:
+                projects_to_test.remove(to_exclude)
+    elif platform == "Windows":
+        for to_exclude in EXCLUDE_WINDOWS:
+            if to_exclude in projects_to_test:
+                projects_to_test.remove(to_exclude)
+    elif platform == "Darwin":
+        for to_exclude in EXCLUDE_MAC:
+            if to_exclude in projects_to_test:
+                projects_to_test.remove(to_exclude)
+    else:
+        raise ValueError("Unexpected platform.")
+    return projects_to_test
+
+
+def _compute_projects_to_build(projects_to_test: Set[str]) -> Set[str]:
+    return _add_dependencies(projects_to_test)
+
+
+def _compute_project_check_targets(projects_to_test: Set[str]) -> Set[str]:
+    check_targets = set()
+    for project_to_test in projects_to_test:
+        if project_to_test not in PROJECT_CHECK_TARGETS:
+            continue
+        check_targets.add(PROJECT_CHECK_TARGETS[project_to_test])
+    return check_targets
+
+
+def _compute_runtimes_to_test(projects_to_test: Set[str]) -> Set[str]:
+    runtimes_to_test = set()
+    for project_to_test in projects_to_test:
+        if project_to_test not in DEPENDENT_RUNTIMES_TO_TEST:
+            continue
+        runtimes_to_test.update(DEPENDENT_RUNTIMES_TO_TEST[project_to_test])
+    return runtimes_to_test
+
+
+def _compute_runtime_check_targets(runtimes_to_test: Set[str]) -> Set[str]:
+    check_targets = set()
+    for runtime_to_test in runtimes_to_test:
+        check_targets.add(PROJECT_CHECK_TARGETS[runtime_to_test])
+    return check_targets
+
+
+def _get_modified_projects(modified_files: list[str]) -> Set[str]:
+    modified_projects = set()
+    for modified_file in modified_files:
+        modified_projects.add(pathlib.Path(modified_file).parts[0])
+    return modified_projects
+
+
+def get_env_variables(modified_files: list[str], platform: str) -> Set[str]:
+    modified_projects = _get_modified_projects(modified_files)
+    projects_to_test = _compute_projects_to_test(modified_projects, platform)
+    projects_to_build = _compute_projects_to_build(projects_to_test)
+    projects_check_targets = _compute_project_check_targets(projects_to_test)
+    runtimes_to_test = _compute_runtimes_to_test(projects_to_test)
+    runtimes_check_targets = _compute_runtime_check_targets(runtimes_to_test)
+    return {
+        "projects_to_build": ";".join(sorted(projects_to_build)),
+        "project_check_targets": " ".join(sorted(projects_check_targets)),
+        "runtimes_to_build": ";".join(sorted(runtimes_to_test)),
+        "runtimes_check_targets": " ".join(sorted(runtimes_check_targets)),
```

Comment:
Why some are `;` separated while other are space-separated?

Reply:
`;` is used as separator in CMake list arguments, while space is used when a list is passed to a shell invocation.

Reply:
Thanks! I suspected something like that. I'd prefer to have this script output be standard, and then do a replace closer to the use location (like in the `generate-buildkite-pipeline-premerge` script)

Reply:
What is the benefit of doing the replace closer to the location? To me, it seems like it just adds (slight) unnecessary complexity on the workflow side without any actual benefit. I think it's reasonably clear that `projects_to_build` and `runtimes_to_build` are going to be passed as CMake arguments to `LLVM_ENABLE_PROJECTS` and `LLVM_ENABLE_RUNTIMES` and the `*_check_targets` variables are going to be passed to a ninja invocation.

I added a comment though which will hopefully make things a bit more clear.

Reply:
A comment is fine.

> I think it's reasonably clear that projects_to_build and runtimes_to_build are going to be passed as CMake arguments

When you know how the CI works. If you don't you see a scripts which purpose is to turn a list of into a list of targets/projects, the use then depends on the caller.